eksconfig/clusterautoscaler.go

```diff
@@ -0,0 +1,56 @@
+package eksconfig
+
+// ClusterAutoscalerSpec defines the spec for the Addon
+type ClusterAutoscalerSpec struct {
+	Image          string        `json:"image,omitempty"`
+	CloudProvider  CloudProvider `json:"cloudProvider,omitempty"`
+	MinNodes       int           `json:"minNodes,omitempty"`
+	MaxNodes       int           `json:"maxNodes,omitempty"`
+	ScaleDownDelay string        `json:"scaleDownDelay,omitempty"`
+}
+
+// CloudProvider enum for ClusterAutoscaler
+type CloudProvider string
+
+// CloudProvider options
+const (
+	CloudProviderKubemark CloudProvider = "kubemark"
+	CloudProviderAWS      CloudProvider = "aws"
+)
+
+// ClusterAutoscalerStatus defines the status for the Addon
+type ClusterAutoscalerStatus struct {
+	AddonStatus `json:"inline"`
+}
+
+// Validate installs the addon
+func (spec *ClusterAutoscalerSpec) Validate(cfg *Config) error {
+	return nil
+}
+
+// Default installs the addon
+func (spec *ClusterAutoscalerSpec) Default(cfg *Config) {
+	if spec.CloudProvider == "" {
+		spec.CloudProvider = CloudProviderKubemark
+	}
+	if spec.MinNodes == 0 {
+		spec.MinNodes = 1
+	}
+	if spec.MaxNodes == 0 {
+		spec.MaxNodes = 100
+	}
+	if spec.ScaleDownDelay == "" {
+		spec.ScaleDownDelay = "30s"
+	}
+	spec.Image = spec.defaultImage()
+}
+
+func (spec *ClusterAutoscalerSpec) defaultImage() string {
+	if spec.Image != "" {
+		return spec.Image
+	}
+	if spec.CloudProvider == CloudProviderKubemark {
+		return "767520670908.dkr.ecr.us-west-2.amazonaws.com/cluster-autoscaler-kubemark:latest"
```

Comment:
For other add-ons, we specify account ID and repo name as separate fields:

https://github.com/aws/aws-k8s-tester/blob/354f0478ba82cf9a1c4fa890eda71956d6ff8c33/eksconfig/add-on-php-apache.go#L23-L33

in order to check if ECR image exists and prevent creating Pod with an invalid image. If it does not exist, we fail fast: before we create Pod and find out later.

What do you think?

I am also fine with this approach since it's simpler.

Reply:
I was leaning away from strong validation. Typically I'd agree with fail fast, I'd only expect users to override image if they know what they're doing, and the additional complexity doesn't seem worth it.